src/query/procedure/module.cpp

```diff
@@ -320,6 +322,368 @@ void RegisterMgTransformations(const std::map<std::string, std::unique_ptr<Modul
   module->AddProcedure("transformations", std::move(procedures));
 }
 
+namespace {
+bool IsAllowedExtension(const auto &extension) {
+  constexpr std::array<std::string_view, 1> allowed_extensions{".py"};
+  return std::any_of(allowed_extensions.begin(), allowed_extensions.end(),
+                     [&](const auto allowed_extension) { return allowed_extension == extension; });
+}
+
+bool IsSubPath(const auto &base, const auto &destination) {
+  const auto relative = std::filesystem::relative(destination, base);
+  return !relative.empty() && *relative.begin() != "..";
+}
+
+std::optional<std::string> ReadFile(const auto &path) {
+  std::ifstream file(path);
+  if (!file.is_open()) {
+    return std::nullopt;
+  }
+
+  const auto size = std::filesystem::file_size(path);
+  std::string content(size, '\0');
+  file.read(content.data(), static_cast<std::streamsize>(size));
+  file.close();
```

Comment:
Does `ifstream` close the file on its destructor? What if there is an exception before the call to close?

Reply:
it seems it does call `close` on destruction :+1: